### PR TITLE
chore(config): Rename `healthcheck` & `healthcheck_uri`

### LINKED ISSUE #4203: Rename `healthcheck` & `healthcheck_uri` to `healthcheck.enabled` & `healthcheck.uri`
Part of our larger configuration cleanup issue.

### PR DESCRIPTION
**Pull request by ktff**
Closes #4203

There are two flaws with the current solution:
* ~~Not fully backward compatible as it will break configurations where both `healthcheck` and `healthcheck_uri` options are used.~~ (EDIT: resolved)
     
* Requires having health check options in outer sink config instead of inner sink config where they belong. This is currently effecting only one option.

### Open questions

- [x] I'm not quite happy with this solution, but besides writing a custom `Deserializer` for `SinkOuter` which seams like too much, what would be better?

cc. @bruceg

### CODE REVIEW

**Comment by bruceg** [comment]
>     * [ ]  I'm not quite happy with this solution, but besides writing a custom `Deserializer` for `SinkOuter` which seams like too much, what would be better?

The other path for compatibility would be to continue to include `healthcheck_uri: Option<UriSerde>` in the configs, and add a `merge_uri` method that is called by all the sinks that use it. The question is if it's worth it compared to breaking old configs using `healthcheck_uri`.

**Comment by bruceg** [comment]
> The other path for compatibility would be to continue to include `healthcheck_uri: Option<UriSerde>` in the configs, and add a `merge_uri` method that is called by all the sinks that use it.

Actually, thinking a little further, could something like this be put into the global `SinkOuter` and then handled for all the sinks transparently? That should restore compatibility without too much fuss.

File: docs/reference/components/sinks/http.cue [hunk 1]
```diff
@@ -114,6 +114,16 @@ components: sinks: http: {
 				examples: ["https://10.22.212.22:9000/endpoint"]
 			}
 		}
+		healthcheck: type: object: options: uri: {
```
Comment (bruceg) [inline docs/reference/components/sinks/http.cue:117]: Would this make more sense as an option to the global `healthcheck` in `sinks.cue`, with a flag to enable it like for `healthcheck.enabled`?

**Comment by ktff** [comment]
True, that would solve it without fuss. I'll add it.

File: src/serde.rs [hunk 2]
```diff
@@ -47,3 +50,51 @@ impl<V: 'static> Fields<V> {
...
     }
 }
+
+/// Enables deserializing from a value that could be a bool or uri or a struct.
+/// Example:
+/// healthcheck: bool
+/// healthcheck: "...."
+/// healthcheck.enabled: bool
+/// Both are accepted.
+pub fn bool_or_str_or_struct<'de, T, D>(deserializer: D) -> Result<T, D::Error>
```
Comment (bruceg) [inline src/serde.rs:60]: If you keep a `healthcheck_uri` option, the serde bits could be simplified to only need a `bool_or_struct` deserializer.

File: docs/reference/components/sinks/http.cue (hunk 1 above)
Comment (ktff) [reply docs/reference/components/sinks/http.cue:117]: It seams useful.

Comment (ktff) [reply docs/reference/components/sinks/http.cue:117]: Only one sink has this option, and there is some work to add this to cue, so let's wait for when it's used in at least two sinks.

Comment (bruceg) [reply docs/reference/components/sinks/http.cue:120]: Agreed

File: src/config/mod.rs [hunk 3]
```diff
@@ -210,19 +214,80 @@ inventory::collect!(SourceDescription);
...
+                .clone()
+                .or_else(|| self.healthcheck_uri.clone()),
+            ..self.healthcheck.clone()
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(default)]
+pub struct HealthcheckOptions {
```
Comment (bruceg) [inline src/config/mod.rs:264]: Rust appears to be complaining about multiple definitions of this, do you know what that's about?

**Review by bruceg — APPROVED** [review]
Looks good except for the compiler error.

File: src/config/mod.rs (hunk 3 above)
Comment (ktff) [reply src/config/mod.rs:264]: A, forgot about that, there was another PR that added a struct with the same name.

**Pull request merged**